girder_worker_utils/transforms/contrib/girder_io.py

```diff
@@ -0,0 +1,68 @@
+import os
+import shutil
+import tempfile
+
+from ..girder_io import GirderClientTransform
+
+
+class GirderFileIdAllowDirect(GirderClientTransform):
+    """
+    This transform either uses direct path to access a file, if possible and
+    allowed, or downloads a Girder File to the local machine and passes its
+    local path into the function.
+
+    WARNING: if a direct path is used, the task MUST NOT modify the file.  It
+    is the resposibility of the user of this transform to ensure tasks treat
+    files as read-only.
+
+    :param _id: The ID of the file to download.
+    :type _id: str
+    """
+    def __init__(self, _id, **kwargs):
+        super(GirderFileIdAllowDirect, self).__init__(**kwargs)
+        from girder.models.file import File
+        from girder.exceptions import FilePathException
+        self.file_id = _id
+        file = File().load(self.file_id, force=True)
```

Comment:
Seeing as this transform module is named `girder_io`, the transform is not intended to be used outside of Girder.  If someone needs that, they can write a different module.  Transforms for Girder should be useful in that context, and not add more burden and repetition than necessary.  I'd rather have this once than in multiple places.

Reply:
Sorry,  using this inside or outside the context of Girder wasn't really the point I was trying to make.  

My primary technical concerns with the PR as-is are: 

1. import guarding inside `__init__` is likely to lead to unexpected behavior and other design problems that increase complexity (e.g. the test mocking).  
2. In many contexts this is going to be called from an environment where the file model is already loaded (e.g. a request handler) and so, as implemented, in those cases there is going to have to be a second trip to the database.

From a more philosophical position: 

3. the File() object is only available in the context of girder-as-application while transforms are intended to be a set of library utilities with minimal dependencies. By tucking imports and the File() object into the `__init__` you're relying on the existence of a request context, a database connection and a whole load of other mechanics. Ultimately it subverts an otherwise clean separation between library and application. 

With regard to (3) I know the `GirderClientTransform` class already breaks this separation.  To be honest I'm increasingly regretting the decision to implement the transform that way, not only because now the API is fixed across a number of projects that are using it,  but also because it appears to have led people to believe that it was a good idea. I am of the opinion that _it was not_ and I want to prevent further spread of the approach. 

The purpose of the transforms are to collect state from the producer and make that state available to the consumer so the task can ignore the I/O details. The initialization of the transform is the mechanism by which that state should be collected.

Reply:
Currently, in the existing girder_io transforms, *each and every*  transform that uses `GirderClientTransform` instantiates a separate instance of the girder client, so if you need two files, you have two clients.   Worse than what you point out in (2), the transform creates an instance of girder client even when it isn't necessary, and then a typical task creates another instance of girder client to return data to girder.

We don't actually want separation and isolation, otherwise we would serialize an entire file and pass that to girder_worker and not use the client at all.  I'd argue we want *less* separation -- to have at most one girder_client when getting multiple files from the same instance of Girder with the same credentials and to reuse that client when returning data to Girder.

As you said, I thought the purpose of the transforms was to collect state from the producer.  I think the other purpose of a transform is to provide utility -- the producer shouldn't have to replicate the same code over and over again if it is always going to be used by the transform.  Do you recommend that the transform be passed the File model and the File document?  Or that there is a separate helper function somewhere else `getDataForGirderFileAllowDirectTransform` which returns the id, name, and path (if available)?

Reply:
Doesn't `GirderFileId` have to depend on `GirderClientTransform`?  And, if, so, doesn't this transform also have to depend on it?  How can it access the file if it can't or shouldn't be reached locally by the worker?

I can't follow you logic between when you think it is useful to have inefficiencies about doing things multiple times versus others -- you objected to the possibility of an extra trip to the database where code avoids having to burden the caller, but are okay with multiple girder clients.  This seems contradictory -- if the goal is to make this clear for the consumer of the transform at the expense of efficiency, that is okay -- but why it is okay in one instance and dunned in another.

Reply:
Round-tripping to the database is a technical concern i have with this PRs current implementation. Maybe I have misunderstood, but it seems unnecessary in the majority of the cases where  someone would actually use this transforms (ie. from a request handler that has already validated/loaded the model with `Description.modelParam`)

Separating the state of transforms from each other on the consumer is an architectural design decision that reduces the complexity of the worker code, prevents developers from having to reason about the order in which various transforms have access to that state, and generally conforms to principles of least surprise. A side effect of this is if you use multiple transforms derived from `GirderClientTransform`  they will have separate girder clients.  It is a choice to be slightly less efficient to improve clarity and robustness. 

None of these points are germane to the overarching issue I have with this PR's current implementation which is that it violates the scope of girder worker transforms by sneaking application dependant behavior into an object that is intended to used as a library.

Reply:
If every time I need to call a transform I have to call a helper function, that helper function should be part of that transform, or there needs to be a wrapper that contains that transform and the helper function.  If this wrapper isn't in this utility library, where should it exist?  It is only needed when using Girder with Girder Worker, and, superficially, this seems like the library that provides utilities for that purpose.

Reply:
> it violates the scope of girder worker transforms by sneaking application dependant behavior into an object that is intended to used as a library.

@kotfic I'm trying to get up to speed on this PR. Can you explain what you mean by application dependent behavior in this case?

Reply:
Just to make sure I understand:

* `__init__` gets called in the request thread
* `transform` gets called on the worker side

Is that right?

Reply:
By moving the import of the `File` model into the `__init__` function and calling `load` and `getLocalFilePath` this transform now depends on the existence of a mongo database and drags in, at run time, a bunch of girder dependencies. Alternately,  an implementation like this:

```python
class GirderFileIdAllowDirect(GirderFileId):

    def __init__(self, _id, file_name=None, local_file_path=None, **kwargs):
        super(GirderFileIdCheckLocal, self).__init__(**kwargs)
        self.file_name = file_name
        self.local_file_path = local_file_path

    # Rest of the class as implemented
```
makes passing in local_file_path and file_name explicit rather than inferred. It means the calling context has to do more work to get these values,  but it prevents the transform from being tightly coupled to girder's code base and a running service somewhere.

Reply:
@zachmullen yeah basically - 

* `__init__` is called in the producer (usually a request thread or an event binding)
* `transform` is called in the consumer (the worker process)

Reply:
Refactor the transform to avoid using the `File` model is fine, but it isn't useful by itself.  We then want a utility function that can be called to get the transform without having to do the same work in each of the callers.  To this end, we could create a `girder_worker_utils.girder_utils` module that would contain something like:
```
from girder.exceptions import FilePathException
from girder.models.file import File
from ..transforms.contrib.girder_io import GirderFileIdAllowDirect


def paramsForGirderFileIdAllowDirectTransform(file):
    try:
        local_path = File().getLocalFilePath(file)
    except FilePathException:
        local_path = None
    return {'_id': str(file['_id']), 'name': file['name'], 'local_path': local_path}


def girderFileAllowDirect(file, **kwargs):
    params = kwargs.copy()
    params.update(paramsForGirderFileIdAllowDirectTransform(file))
    return GirderFileIdAllowDirect(**params)
```

Since the `GirderFileIdAllowDirect` transform is of no use directly, anything using it would instead of doing 
```
from girder_worker_utils.transforms.contrib import GirderFileIdAllowDirect
inputpath=GirderFileIdAllowDirect(file_id)
```
you would do
```
from girder_worker_utils.girder_utils import girderFileAllowDirect
inputpath=girderFileAllowDirect(file)
```

If you don't think utilities to help communicate between Girder and Girder Worker belong in `girder_worker_utils`, what common library do you suggest instead?

Reply:
Well,  I think I've made it clear that I feel the two or three lines of code belongs in the endpoints that are using the transform. if I had to suggest a location where utilities for communication between Girder and Girder Worker belong it would be in  `girder_worker.girder_plugin` package - keeping in mind that that package currently maintains a lot of legacy code that is likely to be removed once the `girder_worker.run` code is removed.

Reply:
While having the code outside of this library reduces the maintenance burden of this library, it increases the cognitive burden of using it and the maintenance burden of all the places that use it.  I disagree with this, but will do so in the interests of actually getting something done.

Also, this is the first place where you've suggested where you think such code aught to exist, but then immediately caution that that location is the wrong place, too.  This suggests the design of this library and girder worker are ill-conceived, as there seems to be no decent way to not replicate code.